lib/services/downloads_service_backend.dart

```diff
@@ -0,0 +1,1515 @@
+
+  @override
+  // This should come from finamp settings if migration needed
+  Future<(String, int)> get storedDatabaseVersion =>
+      Future.value(("FinampIsar", 1));
+
+  @override
+  Future<void> initialize() async {
+    // Isar database gets opened by main
+  }
+
+  Future<void> _store(IsarTaskDataType type, String id, dynamic data) async {
+    type.check(data); // Verify the data object has the correct type
+    String json = jsonEncode(data.toJson());
+    _isar.writeTxnSync(() {
+      _isar.isarTaskDatas
+          .putSync(IsarTaskData(IsarTaskData.getHash(type, id), type, json, 0));
+    });
+  }
+
+  Future<T?> _get<T>(IsarTaskDataType<T> type, String id) async {
+    var item = _isar.isarTaskDatas.getSync(IsarTaskData.getHash(type, id));
+    return (item == null) ? null : type.fromJson(jsonDecode(item.jsonData));
+  }
+
+  Future<List<T>> _getAll<T>(IsarTaskDataType<T> type) async {
+    var items = _isar.isarTaskDatas.where().typeEqualTo(type).findAllSync();
+    return items.map((e) => type.fromJson(jsonDecode(e.jsonData))).toList();
+  }
+
+  Future<void> _remove(IsarTaskDataType type, String? id) async {
+    _isar.writeTxnSync(() {
+      if (id != null) {
+        _isar.isarTaskDatas.deleteSync(IsarTaskData.getHash(type, id));
+      } else {
+        _isar.isarTaskDatas.where().typeEqualTo(type).deleteAllSync();
+      }
+    });
+  }
+}
+
+/// A wrapper for storing various types of download related data in isar as JSON.
+/// Do not confuse the id of this type with the ids that the content types have.
+/// They will not match.
+@collection
+class IsarTaskData<T> {
+  IsarTaskData(this.id, this.type, this.jsonData, this.age);
+  final Id id;
+  String jsonData;
+  @Enumerated(EnumType.ordinal)
+  @Index()
+  final IsarTaskDataType<T> type;
+  // This allows prioritization and uniqueness checking by delete buffer
+  // It is also used as a retry counter by sync buffer.
+  final int age;
+
+  static int globalAge = 0;
+
+  IsarTaskData.build(String stringId, this.type, T data, {int? age})
+      : id = IsarTaskData.getHash(type, stringId),
+        jsonData = _toJson(data),
+        age = age ?? globalAge++;
+
+  static int getHash(IsarTaskDataType type, String id) {
+    return _fastHash(type.name + id);
+  }
+
+  @ignore
+  T get data => type.fromJson(jsonDecode(jsonData));
+  set data(T item) => jsonData = _toJson(item);
+
+  static String _toJson(dynamic item) {
+    switch (item) {
+      case int id:
+        return jsonEncode({"id": id});
+      case (int itemIsarId, bool required, String? viewId):
+        return jsonEncode({
+          "stubId": itemIsarId,
+          "required": required,
+          "view": viewId,
+        });
+      case _:
+        return jsonEncode((item as dynamic).toJson());
+    }
+  }
+
+  /// FNV-1a 64bit hash algorithm optimized for Dart Strings
+  /// Provided by Isar documentation
+  static int _fastHash(String string) {
+    var hash = 0xcbf29ce484222325;
```

Comment:
Is that initial value a salt? If so, we might wanna store it somewhere more "persistent" (as part of the DB version, or even inside the database itself), since we probably need it to read the data inside the DB?

Reply:
I don't know how the hash works, I just copied it from the Isar documentation.  It is used to convert the jellyfin string ids into 64 bit integer ids for use by isar.  It is not cryptographic, and I don't see any reason why the values would ever need to change.

Reply:
Alright, it should be fine then